engine/termops.ml

```diff
@@ -1463,6 +1463,18 @@ let prod_applist sigma c l =
     | _ -> anomaly (Pp.str "Not enough prod's.") in
   app [] c l
 
+let prod_applist_assum sigma n c l =
+  let open EConstr in
+  let rec app n subst c l =
+    if Int.equal n 0 then
+      if l == [] then Vars.substl subst c
+      else anomaly (Pp.str "Not enough arguments.")
+    else match EConstr.kind sigma c, l with
+    | Prod(_,_,c), arg::l -> app (n-1) (arg::subst) c l
+    | LetIn(_,b,_,c), _ -> app (n-1) (Vars.substl subst b::subst) c l
+    | _ -> anomaly (Pp.str "Not enough prod/let's.") in
+  app n [] c l
+
```

Comment:
What's the point of passing an integer to this function? It seems to be completely retrievable from the argument list.

Reply:
Notice that in the `LetIn` case, `n` is decremented but `l` stays the same.
That is, `n` is the length of the removed context, including both `Prod` and `LetIn`, while the length of `l` must be the number of `Prod` in that context.

This is theoretically documented in the comment in `termops.mli` (which I copied from `term.mli`), but admittedly requires a close reading.

Reply:
Right, but the only thing this does in the base case is asserting the length, if that goes wrong it raises an anomaly. So I'd definitely remove the integer from the arguments, this looks like a badly designed API.